proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+.. proposal-number:: 
+
+.. trac-ticket:: 
+
+.. implemented:: 
+
+.. highlight:: haskell
+
+This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/59>`_.
+
+.. contents::
+
+Custom type errors
+==================
+
+This proposes the addition of two combinators over ``Constraint``, namely ::
+
+    IfUndischarged :: Constraint -> ErrorMessage -> Constraint
+    IfApart        :: Type -> Type -> Constraint -> Constraint -> Constraint
+
+which enable library authors to provide custom error messages for their libraries.
+
+In addition, this proposal also motivates changing the constraint solving order to prioritize arguments to functions over the check of the application itself.
+
+This proposal does *not* address whether this facility should be used at all by the libraries bundled with GHC. This might be the topic of another proposal in the future, if this one gets accepted.
+
+Further information (with a more academic flavor) is available in `this draft <http://www.staff.science.uu.nl/~f100183/type-errors-draft.pdf>`_ and this `presentation at Curry On! 2017 <https://www.youtube.com/watch?v=LuqSkWOcnSA>`_.
```

Comment:
I've seen the presentation (in the flesh!), but not the paper. Feel free to refer to sections of the paper in answers to my questions.